### PR TITLE
Multi-part Premium Onboarding

### PR DESCRIPTION
**Pull request by maxxcrawford**
## Summary
This PR adds a new premium onboarding experience. It hides everything else and presents a three-step process to intro the premium features. 

Depends on https://github.com/mozilla-l10n/fx-private-relay-l10n/pull/42

---

⚠️  Tester Notes:  
- You will need to run the following commands/steps to get this working on local:  
    - Run `npm i` (new SCSS source maps)
    - Run `python manage.py migrate` (new database (interger) item storing what step a user completed on onboarding)
    - Add a new ENV variable `MAX_ONBOARDING_AVAILABLE`, set to `3`
- Helper function to reset your current onboarding state: Run the following command in the console: 
```
await patchProfile(document.getElementById("profile-main").dataset.profileId, {onboarding_state: 0});
```

---

Reference Notes:
- [Figma reference file](https://www.figma.com/file/41y1kGrI2zKv4kvfdJdAWW/Relay-Website?node-id=173%3A849)
- Jira cards: [MPP-1299](https://mozilla-hub.atlassian.net/browse/MPP-1299), [MPP-1297](https://mozilla-hub.atlassian.net/browse/MPP-1297), [MPP-1298](https://mozilla-hub.atlassian.net/browse/MPP-1298), [MPP-1296](https://mozilla-hub.atlassian.net/browse/MPP-1296), 
- Added two code-quality fixes: Added sourcemaps (70d0bf72e4517fc3751cd3547fd719cef31df2c5) and updated background color of page to match Figma/Nebula (cb41757b415d8474e6f4b178c9c119da7057fbf8)

---

## TODO
- [ ] Add logic to not promote the add-on link if on mobile/non-Firefox device 
- [x] Add final copy
- [x] Replace hard copy with L10N strings
- [x] Update logic on add-on step to hide second paragraph once add-on is installed
- [x] Add Google Pings for funnel/on boarding completion/skips 
- [x] Update UTMs on add-on download link
- [x] Fetch/Catch: Fail/success state for domain registration
- [x] Replace cookie-based Django messages with DOM/JS-powered logic
- [x] Add database boolean to remember if a user has gone through onboarding 
- [x] Add logic for when to show this to users (First load after premium, etc) and to remember not to show it again.
- [x] Add add-on download link to add-on install CTA

## Testing

Test cases: 
- Free user who upgrades (has aliases created)
- New premium user with no aliases created (brand new premium) and no domain registered
- New premium user with no aliases created (brand new premium) and domain registered (Weird edge case but I solved for it) 
- Any above where the add-on is already installed

Regress tests to also perform: 
- Register domain on dashboard 
- Update label sync preference on settings page

## Screenshots

Video Demo (Taken on f31c3cfad26d15a807adb9cb82ebf24f66c6ef1f): https://drive.google.com/file/d/1lz5i0D_g46dEnrsiJ52hghSAFtEiKRTH/view?usp=sharing

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 const { src, watch, series, dest } = require("gulp");
 const sass = require("gulp-sass")(require("sass"));
 const del = require("del");
+const sourcemaps = require("gulp-sourcemaps");
```
Comment (Vinnl) [inline gulpfile.js:4]: Yes, source maps! :clap:

File: privaterelay/templates/includes/premium-onboarding.html [hunk 2]
```diff
@@ -0,0 +1,127 @@
+{% load static %}
+{% load ftl %}
+{% ftlconf bundle='privaterelay.ftl_bundles.main' %}
+
+<!-- <button class="c-premium-onboarding-dismiss js-premium-onboarding-dismiss"><img src="{% static 'images/x-close.svg' %}" alt="Close"></button> -->
+
+<!-- TODO: Remove/port classes from .c-premium-onboarding  -->
+<div class="c-premium-onboarding c-multipart-premium-onboarding">
+    <ul class="c-premium-onboarding-steps">
```
Comment (Vinnl) [inline privaterelay/templates/includes/premium-onboarding.html:9]: Nit, but I think this should be an ordered lists, since users should be going through the steps in order :) Also for the `<ul>` below.

Comment (Vinnl) [inline privaterelay/templates/includes/premium-onboarding.html:9]: Nit, but I think this should be an ordered lists, since users should be going through the steps in order :)

File: privaterelay/templates/includes/premium-onboarding.html [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                </div>
+                <!-- Step 3 -->
+                <div class="c-premium-onboarding-actions c-premium-onboarding-action-step-3 ">
+                    <button class="c-button-as-text js-premium-onboarding-quit-step">Continue, I’ll download the extension later</button>
+                    <button class="mzp-c-button mzp-t-product js-premium-onboarding-quit-step">Create my first email alias</button>
+                </div>
+            </div>
+    
+
+            <ul class="c-premium-onboarding-progress-bar js-premium-onboarding-progress-bar">
```
Comment (Vinnl) [inline privaterelay/templates/includes/premium-onboarding.html:110]: This should probably be an ordered list too, and there should probably be an accessible representation of what these represent (i.e. steps, possibly already completed) for screen readers.

File: static/scss/partials/onboarding.scss [hunk 4]
```diff
@@ -175,3 +174,235 @@
...
+
+.c-multipart-premium-onboarding {
+    display: none;
+    max-width: $content-lg;
+    margin-left: auto;
+    margin-right: auto;
+    position: relative;
+
+    @media #{$mq-md} {
+        min-height: calc(100vh - 150px);
```
Comment (Vinnl) [inline static/scss/partials/onboarding.scss:202]: Maybe add a comment where the `150px` came from?

File: privaterelay/templates/includes/premium-onboarding.html [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+                <div class="c-premium-onboarding-content">
+                    <div class="c-premium-onboarding-content-image">
+                        <img src="{% static 'images/dashboard-onboarding/man-laptop-email-alt.svg' %}" alt="Man sitting with laptop computer">
+                    </div>
+                    <div class="c-premium-onboarding-content-description">
+                        <p><strong>Use a custom domain for sharing aliases:</strong> With a custom domain, you can make your “newsletter” alias different from your “shopping” alias. </p>
+                        <p>Register your custom domain now</p>
+                        <!-- TODO: Move samp inside domainRegistration form -->
+                        <samp>***@<span>{% ftlmsg 'banner-register-subdomain-example-address' %}</span>.mozmail.com</samp>
```
Comment (Vinnl) [inline privaterelay/templates/includes/premium-onboarding.html:45]: Heh, `<samp>` :raised_hands:

File: static/scss/partials/onboarding.scss [hunk 6]
```diff
@@ -175,3 +174,235 @@
...
+    }
+}
+
+.c-premium-onboarding-action-step-3 .mzp-c-button {
+    display: none;
+}
+
+.c-premium-onboarding-addon-installed {
+    padding: $spacing-md;
+    padding-inline-end: $spacing-lg;
```
Comment (Vinnl) [inline static/scss/partials/onboarding.scss:319]: Wow, how is this the first time I've heard to this property? It's really neat!

File: static/scss/partials/onboarding.scss [hunk 7]
```diff
@@ -175,3 +174,235 @@
...
+}
+
+// Progress Bar
+.c-premium-onboarding-progress {
+    background-color: $color-grey-05;
+    width: 100%;
+    text-align: center;
+    position: sticky;
+    z-index: 2;
+    bottom: 0;
```
Comment (Vinnl) [inline static/scss/partials/onboarding.scss:338]: Hmm, why is this `sticky`? Is that just to make it stick to the bottom? If so, I'd rather wrap the whole thing in a flex container, and make the element above this `flex-grow` > 1, to make the intent match the method. Otherwise this might cause issues in unexpected situations (e.g. maybe mobile screens on landscape, where this might cause overlap?).

File: static/js/domain-registration.js [hunk 8]
```diff
@@ -1,12 +1,16 @@
 (function() {
 	"use strict";
 
+    // Note: The form in which this function is init on is passed across multiple subfunctions (via event.target and 
+    // custom event params). Additionally, based on which form is passed, additional logic/functionality is applied.
+    // The two different forms are:
+    //  1. Multi-step Onboarding (Step 2) / id: "onboardingDomainRegistration"
+    //  2. Dashbaord Page (Banner) / id: "domainRegistrationForm"
```
Comment (Vinnl) [inline static/js/domain-registration.js:8]:
```suggestion
    //  2. Dashboard Page (Banner) / id: "domainRegistrationForm"
```

**Review by Vinnl — COMMENTED** [review]
Code so far looks good to me :+1:

File: privaterelay/templates/includes/premium-onboarding.html (hunk 2 above)
Comment (maxxcrawford) [reply privaterelay/templates/includes/premium-onboarding.html:9]: Ahh! Smart call out.

File: static/scss/partials/onboarding.scss (hunk 4 above)
Comment (maxxcrawford) [reply static/scss/partials/onboarding.scss:202]: 🪄 Magic Number™

File: static/scss/partials/onboarding.scss (hunk 6 above)
Comment (maxxcrawford) [reply static/scss/partials/onboarding.scss:319]: Yeah – Since we talked about start/end for stuff to be more RTL in the future.

File: privaterelay/templates/includes/messages-frontend.html [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{% load ftl %}
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/messages-frontend.html:1]: I preserved the Django message system for other errors, but duplicated it for domain registration since we're rolling our own. 

Note that I'm using the same custom `<fluent…>` element again to expose translated strings to the DOM.

File: privaterelay/templates/includes/modal-domain-registration-confirmation.html [hunk 10]
```diff
@@ -24,12 +24,35 @@ <h3><strong class="js-modal-domain-registration-confirmation-domain-preview"></s
...
+      </button>
       <button type="button" class="mzp-c-button mzp-t-neutral js-modal-domain-registration-cancel">
         {% ftlmsg 'profile-label-cancel' %}
       </button>
-      <button type="submit" disabled class="mzp-c-button mzp-t-product js-modal-domain-registration-submit">
-        {% ftlmsg 'modal-domain-register-button' %}
+    </div>
+  </div>
+
+  <div class="c-modal is-hidden t-domain-registration-success js-domain-registration-success">
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/modal-domain-registration-confirmation.html:36]: This is the "success" state that appears during the onboarding registration process: 
image

_(Note that this is not shown currently if you register from the dashboard/profile page)_

File: privaterelay/templates/includes/premium-onboarding.html [hunk 11]
```diff
@@ -0,0 +1,173 @@
+{% load static %}
+{% load ftl %}
+{% ftlconf bundle='privaterelay.ftl_bundles.main' %}
+
+<!-- <button class="c-premium-onboarding-dismiss js-premium-onboarding-dismiss"><img src="{% static 'images/x-close.svg' %}" alt="Close"></button> -->
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/premium-onboarding.html:5]: TODO: Remove commented code

File: privaterelay/templates/includes/premium-onboarding.html [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+<!-- TODO: Remove/port classes from .c-premium-onboarding  -->
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/premium-onboarding.html:7]: TODO: SCSS Code Cleanup

File: privaterelay/templates/includes/premium-onboarding.html [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+<div data-onboarding-completed-step="{{ user_profile.onboarding_state }}" data-max-onboarding-available="{{ settings.MAX_ONBOARDING_AVAILABLE }}" class="c-premium-onboarding c-multipart-premium-onboarding">
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/premium-onboarding.html:8]: DOM State Storage

File: privaterelay/templates/includes/premium-onboarding.html [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+                        <div class="c-premium-onboarding-action-completed">
+                            <img class="margin-right" src="/static/images/icon-check.svg" alt="">
+                            Relay extension installed!
+                        </div>
+                    </div>
+                </div>
+            </div>
+        </li>
+    </ol>
+    <div class="c-premium-onboarding-progress">
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/premium-onboarding.html:99]: Because of the design (and not wanting to get into a `position:absolute` muck), the action buttons for each step are stored apart from the content of each step. You'll see in the JS how the `is-visible` class gets walked through each list to show the correct step at the right time.

File: privaterelay/templates/includes/premium-onboarding.html [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+                    data-ga="send-ga-pings"
+                    data-event-category="Premium Onboarding"
+                    data-event-label="onboarding-step-1-continue"
+                    data-event-value="1"
+                    data-event-nonInteraction="false"
+                    >Get Started</button>
+                </div>
+                <!-- Step 2 -->
+                <div class="c-premium-onboarding-actions c-premium-onboarding-action-step-2 {% if user_profile.onboarding_state == 1 %} is-visible {% endif %}">
+                    {% if not user_profile.subdomain %}
```
Comment (maxxcrawford) [inline privaterelay/templates/includes/premium-onboarding.html:114]: In Steps 2 and 3, there are two buttons (only one shows at a time): skip and continue once the action is complete

If a user already completed the action on page load, we do not even render the skip button. We still render the continue button regardless of state because if they complete the action, we'll want it.

File: static/js/app.js [hunk 16]
```diff
@@ -663,3 +665,109 @@ faqQuestion.forEach(item => {
   const expandButton = item.querySelector("button");
   expandButton.addEventListener("click", () => showFAQAnswer(item), false);
 });
+
+// TODO: Add element check to init everything.
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply static/js/app.js:669]:
```suggestion
```

File: static/scss/partials/onboarding.scss (hunk 4 above)
Comment (maxxcrawford) [reply static/scss/partials/onboarding.scss:202]:
```suggestion
        // 150px is an approx. size of the header and footer combined
        min-height: calc(100vh - 150px); 
```

File: static/scss/partials/onboarding.scss (hunk 7 above)
Comment (maxxcrawford) [reply static/scss/partials/onboarding.scss:363]: This is sticky as we need the content (onboarding steps/skip buttons) to be visible on screen at all times for mobile users. (Otherwise, it's below the fold and they have to scroll to exit)

**Review by maxxcrawford — COMMENTED** [review]
Big self-review!

File: privaterelay/templates/includes/premium-onboarding.html (hunk 2 above)
Comment (maxxcrawford) [reply privaterelay/templates/includes/premium-onboarding.html]: Revised in e4ab4ee8f06e1a61deaf6de33cafd6be95d0e3de

File: privaterelay/templates/includes/premium-onboarding.html (hunk 3 above)
Comment (maxxcrawford) [reply privaterelay/templates/includes/premium-onboarding.html]: `<ol>` Revised in e4ab4ee8f06e1a61deaf6de33cafd6be95d0e3de

What should I do for screen readers?

File: privaterelay/templates/includes/premium-onboarding.html (hunk 11 above)
Comment (maxxcrawford) [reply privaterelay/templates/includes/premium-onboarding.html:5]:
```suggestion
```

File: static/js/app.js [hunk 17]
```diff
@@ -663,3 +665,109 @@ faqQuestion.forEach(item => {
...
+    }
+
+    const onboardingContainer = document.querySelector(".c-multipart-premium-onboarding");
+    let maxOnboardingState = parseInt(onboardingContainer.dataset.maxOnboardingAvailable, 10);
+    
+    const mainContainer = document.getElementById("profile-main");
+    mainContainer.classList.remove("is-premium-onboarding");
+    
+    const profileId = mainContainer.dataset.profileId;
+    const settings = {
```
Comment (groovecoder) [inline static/js/app.js:756]: `profileData`

File: static/js/app.js [hunk 18]
```diff
@@ -663,3 +665,109 @@ faqQuestion.forEach(item => {
...
+      onboarding_state: maxOnboardingState,
+    };
+
+    try {
+        const response = await patchProfile(profileId, settings);
```
Comment (groovecoder) [inline static/js/app.js:761]: `profileData`

File: static/js/domain-registration.js [hunk 19]
```diff
@@ -54,28 +59,70 @@
             modalSubmit.disabled = false;
             return true;
         },
+        fetchPostSubmit: async ({ domain }) => {
```
[the remark this answers is not in the document]
Comment (groovecoder) [reply static/js/domain-registration.js:62]: yeah this would be good to move over to `api.js` eventually. maybe as simple as a new `postProfileSubdomain` function?

**Review by groovecoder — CHANGES_REQUESTED** [review]

File: static/scss/partials/modals.scss [hunk 20]
```diff
@@ -164,6 +188,27 @@
...
+    .c-modal-header {
+        padding-bottom: $layout-md;
+    }
+    
+    .t-domain-registration-body {
+
+        img {
+            position: relative;
+            margin-top: -$layout-md;
+            margin-bottom: $spacing-md;
```
Comment (codemist) [inline static/scss/partials/modals.scss:202]: We can change this to `$layout-2xs` since its the same value as `$spacing-md`.

File: static/scss/partials/onboarding.scss [hunk 21]
```diff
@@ -175,3 +174,261 @@
...
+.c-multipart-premium-onboarding {
+    display: none;
+    max-width: $content-lg;
+    margin-left: auto;
+    margin-right: auto;
+    position: relative;
+
+    @media #{$mq-md} {
+        // 150px is an approx. size of the header and footer combined
+        min-height: calc(100vh - 150px);
```
Comment (codemist) [inline static/scss/partials/onboarding.scss:203]: 150px is a magic number, what about `$content-xs/2`? `$content-xs` is 304px

File: static/scss/partials/modals.scss (hunk 20 above)
Comment (codemist) [inline static/scss/partials/modals.scss:202]: We can change this to `$layout-2xs` since its the same value as `$spacing-md` (to keep it consistent with the property above it)

File: privaterelay/templates/includes/premium-onboarding.html [hunk 22]
```diff
@@ -30,19 +29,19 @@ <h2>{% ftlmsg 'onboarding-premium-subheadline-html' %}</h2>
...
+                    <h2>{% ftlmsg 'multi-part-onboarding-premium-domain-subheadline' %}</h2>
                 </div>
 
                 <div class="c-premium-onboarding-content">
                     <div class="c-premium-onboarding-content-image">
                         <img src="{% static 'images/dashboard-onboarding/man-laptop-email-alt.svg' %}" alt="Man sitting with laptop computer">
                     </div>
                     <div class="c-premium-onboarding-content-description">
-                        <p><strong>Use a custom domain for sharing aliases:</strong> With a custom domain, you can make your “newsletter” alias different from your “shopping” alias. </p>
+                        <p><strong>{% ftlmsg 'multi-part-onboarding-premium-domain-title' %}</strong>{% ftlmsg 'multi-part-onboarding-premium-domain-description' %}</p>
```
Comment (pdehaan) [inline privaterelay/templates/includes/premium-onboarding.html:41]: Q: Should this have a space after the `</strong>`?

File: static/js/analytics.js [hunk 23]
```diff
@@ -341,7 +341,13 @@ if (!_dntEnabled()) {
...
-      if (isGoogleAnalyticsAvailable()) {
+      const eventNonInteraction = eventData.eventNonInteraction || null;
+      
+      if (eventNonInteraction && isGoogleAnalyticsAvailable()) {
+        ga("send", "event", eventData.eventCategory, eventAction,  eventData.eventLabel, eventValue, {
+          nonInteraction: eventNonInteraction
+        });
+      } else if (isGoogleAnalyticsAvailable()) {
         ga("send", "event", eventData.eventCategory, eventAction,  eventData.eventLabel, eventValue);
       }
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply static/js/analytics.js:352]: This isn't needed at all, actually. I was not scrolling up in the GA debugger – it was running a view event on the _next_ item. All good. Removing my initial code change entirely.

[the remark this answers is not in the document]
Comment (maxxcrawford) [reply static/js/analytics.js]: Removed in a1c3f191dab5eea3763ab1663160911d233cb1c5

File: static/js/app.js [hunk 24]
```diff
@@ -663,3 +665,109 @@ faqQuestion.forEach(item => {
...
+    // Grab data necessary to update settings:onboarding_state value
+    const onboardingContainer = document.querySelector(".c-multipart-premium-onboarding");
+    const mainContainer = document.getElementById("profile-main");
+    const profileId = mainContainer.dataset.profileId;
+    let currentOnboardingState = parseInt(onboardingContainer.dataset.onboardingCompletedStep, 10);
+    
+    // Bump onboarding to next interger
+    currentOnboardingState++
+
+    const settings = {
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply static/js/app.js]: Revised in a1c3f191dab5eea3763ab1663160911d233cb1c5

File: static/js/app.js [hunk 25]
```diff
@@ -663,3 +665,109 @@ faqQuestion.forEach(item => {
...
+    // Update progress bar
+    const activeOnboardingProgressSlides = document.querySelectorAll(".c-premium-onboarding-progress-bar-item.is-completed");
+    // Get most recent "completed" progress bar section
+    const activeOnboardingProgressSlide = activeOnboardingProgressSlides[activeOnboardingProgressSlides.length - 1];
+    activeOnboardingProgressSlide.nextElementSibling.classList.add(
+      "is-completed"
+    );
+
+    try {
+        const response = await patchProfile(profileId, settings);
```
[the remark this answers is not in the document]
Comment (maxxcrawford) [reply static/js/app.js]: Revised in a1c3f191dab5eea3763ab1663160911d233cb1c5

File: static/js/app.js (hunk 17 above)
Comment (maxxcrawford) [reply static/js/app.js]: Revised in a1c3f191dab5eea3763ab1663160911d233cb1c5

File: static/js/app.js (hunk 18 above)
Comment (maxxcrawford) [reply static/js/app.js]: Revised in a1c3f191dab5eea3763ab1663160911d233cb1c5

File: static/scss/partials/modals.scss (hunk 20 above)
Comment (maxxcrawford) [reply static/scss/partials/modals.scss:202]: Oh! actually…`$layout-md` is the same as `$spacing-2xl`. Let's keep it spacing. Adjusted in 8d62d4797b20f3c3d12d995576a0f7728f227368

Comment (maxxcrawford) [reply static/scss/partials/modals.scss:202]: image

File: privaterelay/templates/includes/premium-onboarding.html (hunk 22 above)
Comment (maxxcrawford) [reply privaterelay/templates/includes/premium-onboarding.html:38]:
```suggestion
                        <p><strong>{% ftlmsg 'multi-part-onboarding-premium-domain-title' %}</strong> {% ftlmsg 'multi-part-onboarding-premium-domain-description' %}</p>
```

Comment (maxxcrawford) [reply privaterelay/templates/includes/premium-onboarding.html:38]: Yes it should! :)

**Review by groovecoder — APPROVED** [review]
